### ISSUE TITLE
Does `withgradient` have lower precicision than simply calling the function?

### ISSUE THREAD

**Issue by FelixBenning** [FluxML/Flux.jl#2315]
### Setup

```julia
using Flux:Flux
using MLDatasets: MNIST
using Statistics: Statistics

x_train, y_train = MNIST(split=:train)[:]
x_train = Float32.(x_train)
y_train_oh = Flux.onehotbatch(y_train, 0:9)
size(x_train), size(y_train), size(y_train_oh)


function mnistSimpleCNN7()
	return Flux.Chain(
	    Flux.unsqueeze(3),
	    Flux.Conv((7,7), 1=>48, bias=false), # output becomes 22x22
	    Flux.BatchNorm(48, Flux.relu),
	    Flux.Conv((7,7), 48=>96, bias=false), # output becomes 16x16
	    Flux.BatchNorm(96, Flux.relu),
	    Flux.Conv((7,7), 96=>144, bias=false), # output becomes 10x10
	    Flux.BatchNorm(144, Flux.relu),
	    Flux.Conv((7,7), 144=>192, bias=false), # output becomes 4x4
	    Flux.BatchNorm(192, Flux.relu),
	    Flux.flatten, # results in 4x4x192=3072 dimensions
	    Flux.Dense(3072, 10, bias=false),
	    Flux.BatchNorm(10)
    )
end
```

### Two variants of the same-ish function

```julia
function empiricalMSELossWithGrad(batch_size=10)
	return model -> begin
		indices = rand(1:60000, batch_size)
		return Flux.withgradient(Flux.params(model)) do
			Flux.Losses.mse(model(x_train[:,:,indices]), y_train_oh[:,indices])
		end
	end
end

function empiricalMSELoss(batch_size=10)
	return model -> begin
		indices = rand(1:60000, batch_size)
		return Flux.Losses.mse(model(x_train[:,:,indices]), y_train_oh[:,indices])
	end
end
```

### Different results (unexpected)

```julia
Statistics.var([empiricalMSELoss(1)(mnistSimpleCNN7()) for _ in 1:100])
```
results in something like `2.5722033f-5`. While evaluations using `withgradient`

```julia
Statistics.var([empiricalMSELossWithGrad(1)(mnistSimpleCNN7())[1] for _ in 1:100])
```

results in `2.2428747f-16`

## Suspected Reason

calling `empiricalMSELossWithGrad(1)(mnistSimpleCNN7())[1]` seems to consistently result in `0.1f0`, while the other call varies in output. So it seems to me like `withgradients` must have less precision?

**Comment by ToucheSir** [comment]
Normalization layers behave differently in a gradient context (they auto-switch to "train" mode). If you want an apples-to-apples comparison, use [`testmode!`](https://fluxml.ai/Flux.jl/stable/models/layers/#Flux.testmode!-Tuple{Any,%20Any}) before calling `(with)gradient`.

**Comment by FelixBenning** [comment]
I am assuming that I can't have gradients in testmode? Because I am attempting to do high dimensional Bayesian Optimization, so I need good samples of the loss to estimate covariances. I just wanted the gradients as well to estimate some other parameter. But it seems like I would be better off to do that separately.

**Comment by ToucheSir** [comment]
Yes, you will still have gradients but the BatchNorm stats won't be updating (so the output should be the same as inference). Dropout isn't the only layer with separate train/test behaviour.
